File: edgequake/sdks/java/src/main/java/io/edgequake/sdk/internal/HttpHelper.java
package io.edgequake.sdk.internal;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.edgequake.sdk.EdgeQuakeConfig;
import io.edgequake.sdk.EdgeQuakeException;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;

/**
 * Internal HTTP helper wrapping java.net.http.HttpClient with EdgeQuake auth headers.
 *
 * WHY: Uses java.net.http.HttpClient (built into Java 11+) instead of OkHttp
 * to eliminate external HTTP dependencies. Jackson handles all JSON serialization.
 */
public class HttpHelper {

    private final HttpClient httpClient;
    private final ObjectMapper mapper;
    private final EdgeQuakeConfig config;

    public HttpHelper(EdgeQuakeConfig config) {
        this.config = config;
        this.mapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(config.timeoutSeconds()))
                .build();
    }

    public ObjectMapper mapper() { return mapper; }

    // ── GET ──────────────────────────────────────────────────────────────

    public <T> T get(String path, Map<String, String> params, Class<T> type) {
        var body = doRequest(buildGet(url(path, params)));
        return deserialize(body, type);
    }

    public <T> T get(String path, Map<String, String> params, TypeReference<T> type) {
        var body = doRequest(buildGet(url(path, params)));
        return deserialize(body, type);
    }

    // OODA-40: Add getRaw for raw string responses (e.g., metrics).
    public String getRaw(String path, Map<String, String> params) {
        return doRequest(buildGet(url(path, params)));
    }

    // ── POST ─────────────────────────────────────────────────────────────

    public <T> T post(String path, Object payload, Class<T> type) {
        var body = doRequest(buildPost(url(path, null), serialize(payload)));
        return deserialize(body, type);
    }

    public void postNoContent(String path, Object payload) {
        doRequest(buildPost(url(path, null), serialize(payload)));
    }

    // OODA-40: Add postRaw for raw string responses (e.g., streaming).
    public String postRaw(String path, Object payload) {
        return doRequest(buildPost(url(path, null), serialize(payload)));
    }

    // ── DELETE ───────────────────────────────────────────────────────────

    /** DELETE that may return a body or empty. Does NOT throw on 204/200. */
    public String delete(String path) {
        return doRequest(buildDelete(url(path, null)));
    }

    public <T> T delete(String path, Class<T> type) {
        var body = doRequest(buildDelete(url(path, null)));
        if (body == null || body.isBlank()) return null;
        return deserialize(body, type);
    }

    // ── PATCH ────────────────────────────────────────────────────────────

    public void patch(String path, Object payload) {
        doRequest(buildPatch(url(path, null), serialize(payload)));
    }

    // OODA-40: Added patch with response type.
    public <T> T patch(String path, Object payload, Class<T> type) {
        var body = doRequest(buildPatch(url(path, null), serialize(payload)));
        return deserialize(body, type);
    }

    // ── PUT ──────────────────────────────────────────────────────────────

    // OODA-40: Added PUT method.
    public <T> T put(String path, Object payload, Class<T> type) {
        var body = doRequest(buildPut(url(path, null), serialize(payload)));
        return deserialize(body, type);
    }

    // ── Internal ─────────────────────────────────────────────────────────

    private String url(String path, Map<String, String> params) {
        var sb = new StringBuilder(config.baseUrl());
        sb.append(path);
        if (params != null && !params.isEmpty()) {
            sb.append('?');
            var first = true;
            for (var entry : params.entrySet()) {
                if (!first) sb.append('&');
                sb.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8));
                sb.append('=');
                sb.append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8));
                first = false;
            }
        }
        return sb.toString();
    }

    private HttpRequest.Builder baseRequest(String url) {
        var builder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(config.timeoutSeconds()));

        // WHY: EdgeQuake API uses X-API-Key header (not Authorization: Bearer).
        if (config.apiKey() != null) {
            builder.header("X-API-Key", config.apiKey());
        }
        if (config.tenantId() != null) {
            builder.header("X-Tenant-ID", config.tenantId());
        }
        if (config.userId() != null) {
            builder.header("X-User-ID", config.userId());
        }
        if (config.workspaceId() != null) {
            builder.header("X-Workspace-ID", config.workspaceId());
        }
        return builder;
    }

    private HttpRequest buildGet(String url) {
        return baseRequest(url).GET().build();
    }

    private HttpRequest buildPost(String url, String json) {
        return baseRequest(url)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json != null ? json : "{}"))
                .build();
    }

    private HttpRequest buildDelete(String url) {
        return baseRequest(url).DELETE().build();
    }

    private HttpRequest buildPatch(String url, String json) {
        return baseRequest(url)
                .header("Content-Type", "application/json")
                .method("PATCH", HttpRequest.BodyPublishers.ofString(json != null ? json : "{}"))
                .build();
    }

    // OODA-40: Added buildPut method.
    private HttpRequest buildPut(String url, String json) {
        return baseRequest(url)
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(json != null ? json : "{}"))
                .build();
    }

    private String doRequest(HttpRequest request) {
        try {
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            var status = response.statusCode();
            if (status >= 200 && status < 300) {
                return response.body();
            }
            throw new EdgeQuakeException(status, response.body());
        } catch (EdgeQuakeException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            if (e instanceof InterruptedException) Thread.currentThread().interrupt();
            throw new EdgeQuakeException("HTTP request failed: " + e.getMessage(), e);
        }
    }

    private String serialize(Object payload) {
        if (payload == null) return null;
        try {
            return mapper.writeValueAsString(payload);
        } catch (Exception e) {
            throw new EdgeQuakeException("JSON serialization failed: " + e.getMessage(), e);
        }
    }

    private <T> T deserialize(String body, Class<T> type) {
        if (body == null || body.isBlank()) return null;
        try {
            return mapper.readValue(body, type);
        } catch (Exception e) {
            throw new EdgeQuakeException("JSON deserialization failed: " + e.getMessage(), e);
        }
    }

    private <T> T deserialize(String body, TypeReference<T> type) {
        if (body == null || body.isBlank()) return null;
        try {
            return mapper.readValue(body, type);
        } catch (Exception e) {
            throw new EdgeQuakeException("JSON deserialization failed: " + e.getMessage(), e);
        }
    }
}

File: edgequake/sdks/kotlin/src/main/kotlin/io/edgequake/sdk/internal/HttpHelper.kt
package io.edgequake.sdk.internal

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.edgequake.sdk.EdgeQuakeConfig
import io.edgequake.sdk.EdgeQuakeException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * Internal HTTP helper using java.net.http.HttpClient.
 * WHY: Zero external HTTP dependencies — JDK 11+ built-in.
 */
open class HttpHelper(@PublishedApi internal val config: EdgeQuakeConfig) {

    @PublishedApi internal val mapper = jacksonObjectMapper().apply {
        configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        setSerializationInclusion(JsonInclude.Include.NON_NULL)
    }

    @PublishedApi internal val httpClient: HttpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(config.timeoutSeconds))
        .build()

    // ── Public API ───────────────────────────────────────────────────

    inline fun <reified T> get(path: String): T =
        execute(buildRequest(path, "GET", null))

    inline fun <reified T> post(path: String, body: Any? = null): T =
        execute(buildRequest(path, "POST", body))

    inline fun <reified T> put(path: String, body: Any? = null): T =
        execute(buildRequest(path, "PUT", body))

    inline fun <reified T> patch(path: String, body: Any? = null): T =
        execute(buildRequest(path, "PATCH", body))

    inline fun <reified T> delete(path: String): T =
        execute(buildRequest(path, "DELETE", null))

    open fun deleteRaw(path: String): String {
        val req = buildRequest(path, "DELETE", null)
        val resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() !in 200..299) {
            throw EdgeQuakeException(
                "HTTP ${resp.statusCode()}: ${resp.body()}",
                resp.statusCode(), resp.body()
            )
        }
        return resp.body()
    }

    open fun getRaw(path: String): String {
        val req = buildRequest(path, "GET", null)
        val resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() !in 200..299) {
            throw EdgeQuakeException(
                "HTTP ${resp.statusCode()}: ${resp.body()}",
                resp.statusCode(), resp.body()
            )
        }
        return resp.body()
    }

    open fun postRaw(path: String, body: Any? = null): String {
        val req = buildRequest(path, "POST", body)
        val resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() !in 200..299) {
            throw EdgeQuakeException(
                "HTTP ${resp.statusCode()}: ${resp.body()}",
                resp.statusCode(), resp.body()
            )
        }
        return resp.body()
    }

    // ── Internals ────────────────────────────────────────────────────

    fun buildRequest(path: String, method: String, body: Any?): HttpRequest {
        val url = "${config.baseUrl}$path"
        val builder = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .timeout(Duration.ofSeconds(config.timeoutSeconds))
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")

        config.apiKey?.let { builder.header("X-API-Key", it) }
        config.tenantId?.let { builder.header("X-Tenant-ID", it) }
        config.userId?.let { builder.header("X-User-ID", it) }
        config.workspaceId?.let { builder.header("X-Workspace-ID", it) }

        val bodyPublisher = if (body != null) {
            HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body))
        } else if (method == "POST" || method == "PUT" || method == "PATCH") {
            HttpRequest.BodyPublishers.ofString("{}")
        } else {
            HttpRequest.BodyPublishers.noBody()
        }

        builder.method(method, bodyPublisher)
        return builder.build()
    }

    inline fun <reified T> execute(request: HttpRequest): T {
        try {
            val resp = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            if (resp.statusCode() !in 200..299) {
                throw EdgeQuakeException(
                    "HTTP ${resp.statusCode()}: ${resp.body()}",
                    resp.statusCode(), resp.body()
                )
            }
            // WHY: 204 No Content returns empty body — handle gracefully
            val body = resp.body()
            if (body.isNullOrBlank()) {
                @Suppress("UNCHECKED_CAST")
                return when {
                    T::class == Unit::class -> Unit as T
                    T::class == Map::class -> emptyMap<String, Any?>() as T
                    T::class == String::class -> "" as T
                    else -> throw EdgeQuakeException("Empty response body for type ${T::class}")
                }
            }
            return mapper.readValue(body, object : TypeReference<T>() {})
        } catch (e: EdgeQuakeException) {
            throw e
        } catch (e: Exception) {
            throw EdgeQuakeException("Request failed: ${e.message}", cause = e)
        }
    }
}
